### PR TITLE
Add code to handle teardown errors and disruptive tests - IBM Power

### PR DESCRIPTION
**Pull request by gitsridhar**
These changes add to the platform nodes framework to fix teardown
errors and relevant disruptive tests

Signed-off-by: Sridhar Venkat <svenkat@MacBook-Pro-4.attlocal.net>

I have tested the code and removed WIP from the title.

### CODE REVIEW

**Comment by gitsridhar** [comment]
I am using the following tier4a test case to validate this code:
run-ci -m "tier4a and manage" --ocsci-conf conf/ocsci/production_powervs_upi.yaml --cluster-name ocstest --cluster-path ~ tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py::TestAddCapacityNodeRestart::test_add_capacity_node_restart[1-workload_storageutilization_rbd0]

File: ocs_ci/ocs/platform_nodes.py [hunk 1]
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
         return default_config_dict
 
 
+class IBMPowerNodes(BaremetalNodes):
```
Comment (zmc) [inline ocs_ci/ocs/platform_nodes.py:1689]: I'm wondering about the naming and inheritance here. I can't see anything specific to Power, and it seems to me this could be useful for anyone doing an libvirt deployment. Perhaps it ought to be `LibvirtNodes`?

File: ocs_ci/ocs/platform_nodes.py [hunk 2]
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
...
+    """
+    IBM Power Nodes class
+    """
+
+    def ___init__(self):
+        super(IBMPowerNodes, self).__init__()
+        self.powernodes = powernodes.POWERNODES()
```
Comment (zmc) [inline ocs_ci/ocs/platform_nodes.py:1696]: I don't understand the need for the separate class here; I'd argue that all functionality belongs in one class (even if it should potentially be in a different module and imported into `power_nodes`)

File: ocs_ci/utility/powernodes.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+class POWERNODES(object):
+    """
+    Wrapper for PowerNodes
+    """
+
+    def __init__(self):
+        """
+        Future use
+        """
+        self.return_values = ""
```
Comment (zmc) [inline ocs_ci/utility/powernodes.py:21]: If this method isn't doing anything right now, I'd leave it out

**Review by zmc — CHANGES_REQUESTED** [review]
Nice work! I appreciate the docstrings. I have some comments, and also a question:

How might this affect deployments done with ocs-upi-kvm?

File: ocs_ci/ocs/platform_nodes.py (hunk 1 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1689]: I put together these changes with the future in mind, for Power currently do things on libvirt, but we certainly would like to do these in PowerVS (IBM Cloud) and Power Baremetal as well one day. When we do that, we may reuse part of Baremetal code that already exists.

File: ocs_ci/ocs/platform_nodes.py (hunk 2 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1696]: I see a pattern among other players in this code, they drive changes to OCP from platform_nodes and underlying code is from ocs_ci.utility. I wanted to keep the same structure for this new code as well.

File: ocs_ci/ocs/platform_nodes.py (hunk 1 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1689]: I put together these changes with the future in mind, for Power currently does things on libvirt, but we certainly would like to do these in PowerVS (IBM Cloud) and Power Baremetal as well one day. When we do that, we may reuse part of Baremetal code that already exists.

File: ocs_ci/ocs/platform_nodes.py (hunk 2 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1696]: I see a pattern, changes to OCP infrastructure are driven from platform_nodes, but underlying changes are maintained in utility. I wanted to keep the same structure for this new code.

Comment (lukebrowning) [reply ocs_ci/ocs/platform_nodes.py:1696]: > I don't understand the need for the separate class here; I'd argue that all functionality belongs in one class (even if it should potentially be in a different module and imported into `power_nodes`)

Zack, I am not sure I understand your comment.   There are multiple classes today and none of them support libvirt or powervs.   Power actually has two platforms - libvirt and powervs.   There is no standardization around libvirt as each node may have a different VM name.   Power follows this pattern test-<ocp version>-<role>, but another implementation may utilize a different naming scheme.  This is really specific to IBM.  There is no platform independent implementation for libvirt running ocs-ci today, just IBM.   That being said, this code would probably result in an exception on powervs as libvirt is not even configured.  The virsh command is probably not even installed.  So, I think the new class for power needs to import modules for ibm-libvirt and powervs.

Comment (zmc) [reply ocs_ci/ocs/platform_nodes.py:1696]: If this is just following existing patterns in the project, I'll just withdraw my comment.

**Review by zmc — APPROVED** [review]

**Comment by zmc** [comment]
I see this is still marked as a WIP; is it still being tested or was that just a reflection of my previous review?

File: ocs_ci/ocs/platform_nodes.py [hunk 4]
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
         return default_config_dict
 
 
+class IBMPowerNodes(NodesBase):
```
Comment (lukebrowning) [inline ocs_ci/ocs/platform_nodes.py:1689]: Power has two platforms - libvirt and powervs.   This class should be updated to import modules to implement callouts for these two platforms.   libvirt sounds like it mght be standardized, but it is not really.   Each implementation may choose different names.   Power follows this pattern test-ocp-4-6-worker-0, but another implementation may utilize a different naming scheme.   On power, we may implement helper functions that invoke virsh or other functions.   Anyway, none of this would apply to other libvirt platforms as only power implements libvirt ocs-ci.   The powervs module should return not implemented for each callout.  Later, we can plug these with the proper callouts.

File: ocs_ci/framework/pytest_customization/marks.py [hunk 5]
```diff
@@ -176,7 +176,10 @@
 )
 
 skipif_bm = pytest.mark.skipif(
-    config.ENV_DATA["platform"].lower() == "baremetal"
+    (
+        config.ENV_DATA["platform"].lower() == "baremetal"
+        or config.ENV_DATA["platform"].lower() == "powervs"
```
[the remark this answers is not in the document]
Comment (gitsridhar) [reply ocs_ci/framework/pytest_customization/marks.py:181]: Luke, I will remove this decorator modification. We may not need this and if required I will create a new PR.

File: ocs_ci/ocs/platform_nodes.py (hunk 4 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1692]: Sure. For now, I am focussing on the libvirt environment. We can take up the PowerVS environment in a separate PR.

File: ocs_ci/framework/pytest_customization/marks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukebrowning) [reply ocs_ci/framework/pytest_customization/marks.py:181]: Did you forget to push the change that removes the decorator?

File: ocs_ci/ocs/platform_nodes.py (hunk 4 above)
Comment (lukebrowning) [reply ocs_ci/ocs/platform_nodes.py:1692]: The code still seems like it would try to invoke virsh on powervs.   What am I missing?

File: ocs_ci/utility/powernodes.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        Start PowerNode Machines
+
+        Args:
+            powernode_machines (list): List of PowerNode machines
+            wait (bool): Wait for PowerNodes to start - for future use
+
+        """
+        for node in powernode_machines:
+            result = exec_cmd(f"sudo virsh start test-ocp4-6-{node.name}")
```
Comment (lukebrowning) [inline ocs_ci/utility/powernodes.py:78]: You hardcoded the ocp version in the string.   This won't work for other ocp versions.

**Review by lukebrowning — CHANGES_REQUESTED** [review]
Sridhar, you marked my comments as resolved.   Did you forget to push the code changes that resolved them.   You agreed to eliminate the change to the decorator, but the change is still in the latest commit.   The second comment is that your changes donot apply to powervs, but there is nothing to prohibit virsh commands from being invoked on powervs.   These commands are not installed on powervs.   It would be better to have the existing message which is that the function is not implemented, rather than spurious errors like command not found.

File: ocs_ci/utility/powernodes.py (hunk 6 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: I have removed the hardcoded OCP version now in the latest update.

**Comment by gitsridhar** [comment]
I have added code: 
1. to check for the KVM environment and execute the stop/start/restart functionality. In other environments where virsh is not possible, the code returns an 'unimplemented' exception.
2. Removed hardcoding of OCP version 4-6 and replaced it with a utility API call.

This is in WIP and more testings, I will remove WIP from the title, and the review can be completed.

**Comment by gitsridhar** [comment]
Used run-ci -m "tier4a and manage" --ocsci-conf conf/ocsci/production_powervs_upi.yaml --cluster-name ocstest --cluster-path ~ tests/manage/z_cluster/nodes/test_nodes_restart.py::TestNodesRestart::test_nodes_restart[True] 
to test this work.

**Comment by gitsridhar** [comment]
I have removed the WIP tag from the title. Please perform a review. @zmc @petr-balogh @vasukulkarni @lukebrowning

**Comment by gitsridhar** [comment]
I just published another version - I added IBM Power Platform to ON_PREM in the constants file. I removed 'powervs' from the bare-metal decorator as well.

**Comment by gitsridhar** [comment]
@zmc @petr-balogh @vasukulkarni @lukebrowning I made a change based on Luke's comment (move lsmod execution to init method). Please review and merge this code.

I also see Luke's comment asking me to move lsmod to init method is missing now. Strange.

**Comment by lsmcfadden** [comment]
@gitsridhar does "I also see Luke's comment asking me to move lsmod to init method is missing now." mean you intend to edit the PR again?

@zmc @petr-balogh @vasukulkarni can you review?  Are you okay to merge it?

**Review by lukebrowning — COMMENTED** [review]
Looks good to me.   I have tested this over all of the tier tests.

**Review by vasukulkarni — APPROVED** [review]

File: ocs_ci/framework/pytest_customization/marks.py [hunk 7]
```diff
@@ -176,7 +176,7 @@
 )
 
 skipif_bm = pytest.mark.skipif(
-    config.ENV_DATA["platform"].lower() == "baremetal"
+    (config.ENV_DATA["platform"].lower() == "baremetal")
```
Comment (vasukulkarni) [inline ocs_ci/framework/pytest_customization/marks.py:179]: why was this changed added?

Comment (gitsridhar) [reply ocs_ci/framework/pytest_customization/marks.py:179]: Vasu, I added code and removed it, so the braces stayed. Let me remove the braces.

**Review by lukebrowning — DISMISSED** [review]

File: ocs_ci/utility/powernodes.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+from ocs_ci.ocs.exceptions import UnexpectedBehaviour
+from ocs_ci.ocs.node import wait_for_nodes_status, get_worker_nodes, get_master_nodes
+from ocs_ci.ocs.ocp import wait_for_cluster_connectivity
+from ocs_ci.utility.utils import TimeoutSampler, exec_cmd
+from ocs_ci.utility.utils import get_ocp_version
+
+logger = logging.getLogger(__name__)
+
+
+class POWERNODES(object):
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:13]: Class names should be PascalCase, not all caps.

```suggestion
class PowerNodes(object):
```

File: ocs_ci/ocs/platform_nodes.py [hunk 9]
```diff
@@ -1712,6 +1713,102 @@ def read_default_config(self, default_config_path):
...
+
+        """
+        if self.powernodes.iskvm():
+            self.powernodes.start_powernodes_machines(nodes, force=force)
+        else:
+            raise NotImplementedError(
+                "This is not libvirt environment. Start nodes not implemented"
+            )
+
+    def restart_nodes(self, nodes, timeout=540, wait=True, force=True):
```
Comment (clacroix12) [inline ocs_ci/ocs/platform_nodes.py:1757]: `timeout` and `wait` args are not used or included in the docstring.

File: ocs_ci/utility/powernodes.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    Wrapper for PowerNodes
+    """
+
+    isKVM = False
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:18]: Does this need to be defined here or can we stick to initializing in the `__init__`?

File: ocs_ci/utility/powernodes.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        else:
+            self.isKVM = False
+        logger.info(f"iskvm check: {self.isKVM}")
+
+    def iskvm(self):
+        """
+        Verify if this environment is libvirt or kvm.
+        Args: None
+        Returns:
+            bool: True if this is kvm environment, False otherwise
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:38]:
```suggestion
        Verify if this environment is libvirt or kvm.
        
        Returns:
            bool: True if this is kvm environment, False otherwise

```

File: ocs_ci/utility/powernodes.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        wait_for_nodes_status(
+            node_names=get_master_nodes(), status=constants.NODE_READY, timeout=900
+        )
+        wait_for_nodes_status(
+            node_names=get_worker_nodes(), status=constants.NODE_READY, timeout=900
+        )
+
+    def restart_powernodes_machines(self, powernode_machines, force=True):
+        """
+
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:122]: No need for the empty line at the start of the docstring

```suggestion
```

File: ocs_ci/utility/powernodes.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+            force (bool): True for PowerNode ungraceful power off, False for
+                graceful PowerNode shutdown - for future use
+
+        Raises:
+            UnexpectedBehaviour: If PowerNode machine is still up
+
+        """
+        ocpversion = get_ocp_version("-")
+        for node in powernode_machines:
+            cmd = "sudo virsh shutdown test-ocp" + ocpversion + f"-{node.name}"
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:76]: We shouldn't be using concatenation and f-strings together like this. Let's simplify and make the whole thing an f-string.
 
```suggestion
            cmd = f"sudo virsh shutdown test-ocp{ocpversion}-{node.name}"
```

File: ocs_ci/utility/powernodes.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+            result = exec_cmd(cmd)
+            """
+            if result.returncode == 0:
+                return False
+            else
+                return True
+            """
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:83]: Does this belong here?

File: ocs_ci/utility/powernodes.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+        Args:
+            powernode_machines (list): List of PowerNode machines
+            wait (bool): Wait for PowerNodes to start - for future use
+
+        """
+        ocpversion = get_ocp_version("-")
+        for node in powernode_machines:
+            result = exec_cmd(
+                "sudo virsh start test-ocp" + ocpversion + f"-{node.name}"
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:108]: Same thing here, just use one f-string:

```suggestion
                f"sudo virsh start test-ocp{ocpversion}-{node.name}"
```

File: ocs_ci/utility/powernodes.py (hunk 8 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Completed in the latest push.

File: ocs_ci/ocs/platform_nodes.py (hunk 9 above)
Comment (gitsridhar) [reply ocs_ci/ocs/platform_nodes.py:1761]: Completed with latest push.

File: ocs_ci/utility/powernodes.py (hunk 10 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Completed with the latest push. Removed global entry.

File: ocs_ci/utility/powernodes.py (hunk 11 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Done.

File: ocs_ci/utility/powernodes.py (hunk 13 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Done.

File: ocs_ci/utility/powernodes.py (hunk 14 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Removed commented code.

File: ocs_ci/utility/powernodes.py (hunk 15 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Done.

File: ocs_ci/utility/powernodes.py (hunk 12 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py]: Done.

**Comment by gitsridhar** [comment]
@clacroix12 I took care of your review comments, after more tests, I will remote WIP tag.

**Comment by gitsridhar** [comment]
Test complete, I removed WIP from the title. Please review.

**Review by vasukulkarni — APPROVED** [review]

File: ocs_ci/utility/powernodes.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+from ocs_ci.ocs.exceptions import UnexpectedBehaviour
+from ocs_ci.ocs.node import wait_for_nodes_status, get_worker_nodes, get_master_nodes
+from ocs_ci.ocs.ocp import wait_for_cluster_connectivity
+from ocs_ci.utility.utils import TimeoutSampler, exec_cmd
+from ocs_ci.utility.utils import get_ocp_version
+
+logger = logging.getLogger(__name__)
+
+
+class POWERNodes(object):
```
Comment (clacroix12) [inline ocs_ci/utility/powernodes.py:13]: Must have missed this one in my initial review. This class should be `PowerNodes`.

**Review by clacroix12 — APPROVED** [review]
Just the one issue that can be taken up in a follow-up PR.

**Comment by lukebrowning** [comment]
@vasukulkarni Please merge this.

File: ocs_ci/utility/powernodes.py (hunk 16 above)
Comment (gitsridhar) [reply ocs_ci/utility/powernodes.py:13]: Sure. I will submit a new PR later to rename this class.
